actioncable/lib/action_cable/connection/base.rb

```diff
@@ -120,19 +127,29 @@ def send_async(method, *arguments)
         worker_pool.async_invoke(self, method, *arguments)
       end
 
-      # Return a basic hash of statistics for the connection keyed with <tt>identifier</tt>, <tt>started_at</tt>, <tt>subscriptions</tt>, and <tt>request_id</tt>.
+      # Return a basic hash of statistics for the connection keyed with <tt>identifier</tt>, <tt>started_at</tt>,
+      # <tt>subscriptions</tt>, <tt>request_id</tt>, and <tt>last_message_received_at</tt>.
       # This can be returned by a health check against the connection.
       def statistics
         {
           identifier: connection_identifier,
           started_at: @started_at,
           subscriptions: subscriptions.identifiers,
-          request_id: @env["action_dispatch.request_id"]
+          request_id: @env["action_dispatch.request_id"],
+          last_message_received_at: @last_message_received_at
         }
       end
 
       def beat
-        transmit type: ActionCable::INTERNAL[:message_types][:ping], message: Time.now.to_i
+        unless @expects_pong
+          return transmit(type: ActionCable::INTERNAL[:message_types][:ping], message: Time.now.to_i)
+        end
+
+        if missed_pong_response?
+          return close(reason: ::ActionCable::INTERNAL[:disconnect_reasons][:heartbeat_timeout])
```

Comment:
I just found an interesting edge case.
If the connection protocol is Hybi (`WebSocket::Driver::Hybi`) and the client abruptly disconnects, calling `#close` on the websocket won't trigger `#on_close` because [the driver handles that case differently](https://github.com/faye/websocket-driver-ruby/blob/main/lib/websocket/driver/hybi.rb#L150-L154).

The end result is that the connection lingers:
```
13:24:46 web.1  | I, [2023-12-08T13:24:46.507291 #440625]  INFO -- : [LATENCY] [device:Device-586795065] 0.008601ms                                                                                                                                                                 
13:24:49 web.1  | I, [2023-12-08T13:24:49.505572 #440625]  INFO -- : [LATENCY] [device:Device-586795065] 0.006799ms                                                                                                                                                                 
13:24:58 web.1  | I, [2023-12-08T13:24:58.499496 #440625]  INFO -- : MISSED PONG                                                                                                                                                                                                    
13:24:58 web.1  | I, [2023-12-08T13:24:58.499582 #440625]  INFO -- : Called Connection#close with reason: heartbeat_timeout and reconnect: true                                                                                                                                     
13:24:58 web.1  | I, [2023-12-08T13:24:58.499659 #440625]  INFO -- : Called WebSocket#close                                                                                                                                                                                         
13:24:58 web.1  | I, [2023-12-08T13:24:58.499679 #440625]  INFO -- : Called SocketClient#close with code:  and reason:                                                                                                                                                              
13:24:58 web.1  | I, [2023-12-08T13:24:58.502117 #440625]  INFO -- : Called SocketClient#close
13:25:01 web.1  | I, [2023-12-08T13:25:01.499804 #440625]  INFO -- : MISSED PONG
13:25:01 web.1  | I, [2023-12-08T13:25:01.499849 #440625]  INFO -- : Called Connection#close with reason: heartbeat_timeout and reconnect: true
13:25:01 web.1  | I, [2023-12-08T13:25:01.499926 #440625]  INFO -- : Called WebSocket#close
13:25:01 web.1  | I, [2023-12-08T13:25:01.499940 #440625]  INFO -- : Called SocketClient#close with code:  and reason: 
13:25:01 web.1  | I, [2023-12-08T13:25:01.501111 #440625]  INFO -- : Called SocketClient#close
```